dcpy/library/config.py

```diff
@@ -79,113 +64,87 @@ def version_today(self) -> str:
-                "uid"
-            ]
-            _format = self.parsed_unrendered_template["dataset"]["source"]["socrata"][
-                "format"
-            ]
-            config = self.parsed_rendered_template(version=self.version_socrata(_uid))
-
-            if _format == "csv":
-                url = f"https://data.cityofnewyork.us/api/views/{_uid}/rows.csv"
-            if _format == "geojson":
-                url = f"https://nycopendata.socrata.com/api/geospatial/{_uid}?method=export&format=GeoJSON"
-            if _format == "shapefile":
-                local_path = f"library/tmp/{config['dataset']['name']}.zip"
-                url = f"https://data.cityofnewyork.us/api/geospatial/{_uid}?method=export&format=Shapefile"
+        Validator(yaml.safe_load(self.unparsed_unrendered_template))()
+
+        _config = self.parsed_unrendered_template
+        if _config.source.socrata:
+            version = self.version_socrata(_config.source.socrata.uid)
+        else:
+            version = self.version or _config.version or self.version_today
+        config = self.parsed_rendered_template(version=version)
+
+        # need to be a bit explicit because of truthiness of empty dicts
+        if _config.source.script is not None:
+            module = importlib.import_module(f"dcpy.library.script.{_config.name}")
+            scriptor = module.Scriptor(config=config.model_dump())
+            path = scriptor.runner()
+            config.source.gdalpath = format_url(path)
+
+        elif _config.source.socrata:
+            socrata = _config.source.socrata
+            if socrata.format == "csv":
+                path = f"https://data.cityofnewyork.us/api/views/{socrata.uid}/rows.csv"
+            elif socrata.format == "geojson":
+                path = f"https://nycopendata.socrata.com/api/geospatial/{socrata.uid}?method=export&format=GeoJSON"
+            elif socrata.format == "shapefile":
+                path = f"library/tmp/{config.name}.zip"
+                url = f"https://data.cityofnewyork.us/api/geospatial/{socrata.uid}?method=export&format=Shapefile"
                 os.system("mkdir -p library/tmp")
-                os.system(f'curl -o {local_path} "{url}"')
-                url = local_path
-
-            options = config["dataset"]["source"]["options"]
-            geometry = config["dataset"]["source"]["geometry"]
-            config["dataset"]["source"] = {
-                "url": {"path": url, "subpath": ""},
-                "options": options,
-                "geometry": geometry,
-            }
-
-        path = config["dataset"]["source"]["url"]["path"]
-        subpath = config["dataset"]["source"]["url"]["subpath"]
-        config["dataset"]["source"]["url"]["gdalpath"] = format_url(path, subpath)
-        config["execution_details"] = get_execution_details()
+                os.system(f'curl -o {path} "{url}"')
+            else:
+                raise Exception("Impossible")
```

Comment:
@alexrichey saw you commented so going to respond - pydantic does validate that string belongs to one of the literals listed, and this is to appease mypy. I did mean to at least return a sensible error because this as written is slightly confusing.

Reply:
Thanks - yeah, I commented and then saw that we'd defined these in a literal, and assumed it was a `mypy` fix.